### PR TITLE
[12.0][IMP] kpi_dashboard: Allow to use formula

### PR DESCRIPTION
**Pull request by etobella**
Kpi Dashboard Formula
=====================

Define KPIs for dashboard using python formulas directly.
### Usage

* Access `Dashboards > Configuration > KPI Dashboards > Configure KPI`
* Create a new KPI and choose as computation method `Code`
* Define the code directly. You can use:

   * `self` and `model` as the kpi element
   * You must return result as a dictionary that will be stored as data

### CODE REVIEW

**Review by legalsylvain — COMMENTED** [review]
Thanks !
Lgtm. Code review, no test. 
However, i'm asking why this feature is not in the main module.
It's a,very generic feature, and more modular than the default one.

**Comment by pedrobaeza** [comment]
Yeah, indeed it's not adding any extra dependency in Odoo modules nor libraries, and the base code is not a burden, as it's very few, so I would include it on the main module.

**Comment by etobella** [comment]
Ok, I will include it then :wink:

**Comment by etobella** [comment]
@legalsylvain @pedrobaeza Now it is integrated directly on kpi_dashboard. Easier now.

**Comment by etobella** [comment]
Just a question, with this change I can probably delete the testing module. What do you think?

**Comment by pedrobaeza** [comment]
If possible, of course I prefer to remove any testing module.

File: kpi_dashboard/readme/CONFIGURE.rst [hunk 1]
```diff
@@ -8,6 +8,9 @@ Configure KPIs
    #. Meter: result must contain `value`, `min` and `max`
    #. Graph: result must contain a list on `graphs` containing `values`, `title` and `key`
 
+#. In order to compute the KPI you can use a predefined function from a model or
+   use the code to directly compute it.
```
Comment (pedrobaeza) [inline kpi_dashboard/readme/CONFIGURE.rst:12]: Isn't this missing the explanation of what to return?

**Review by pedrobaeza — APPROVED** [review]

**Review by jarroyomorales — APPROVED** [review]

File: kpi_dashboard/readme/CONFIGURE.rst (hunk 1 above)
Comment (etobella) [reply kpi_dashboard/readme/CONFIGURE.rst:12]: Ups, Sorry, comments attended

File: kpi_dashboard/models/kpi_kpi.py [hunk 2]
```diff
@@ -84,6 +87,30 @@ def write(self, vals):
...
 
+    def _get_code_input_dict(self):
+        return {
+            "self": self,
+            "model": self.browse(),
+        }
+
+    def _compute_value_code(self):
+        new_cr = Registry(self.env.cr.dbname).cursor()
+        try:
```
Comment (etobella) [inline kpi_dashboard/models/kpi_kpi.py]: This tries to disallow something like self.unlink()

We are trying to find a better option...

Comment (pedrobaeza) [reply kpi_dashboard/models/kpi_kpi.py]: Put it inside a savepoint and do rollback after:
```python
self.env.cr.execute("savepoint kpi_formula_%" % self.id)
safe_eval(...)
self.env.cr.execute("rollback to kpi_formula_%" % self.id)
```

Comment (etobella) [reply kpi_dashboard/models/kpi_kpi.py]: Thanks @pedrobaeza your solution is good, however the savepoint is not perfect. I mean, if the user puts 
```
self.unlink()
self.env.cr.commit()
```
the unlink will be performed. However, the same happened with my solution and this is already cleaner, so I will keep it if nothing better can be found.

Comment (pedrobaeza) [reply kpi_dashboard/models/kpi_kpi.py]: Why don't you simply look for `.commit()` in a constraint and forbids that string?

Comment (etobella) [reply kpi_dashboard/models/kpi_kpi.py]: That will probably be the best combined with the savepoint

File: kpi_dashboard/models/kpi_kpi.py [hunk 3]
```diff
@@ -84,6 +88,27 @@ def write(self, vals):
...
         return super().write(vals)
 
+    def _get_code_input_dict(self):
+        return {
+            "self": self,
+            "model": self.browse(),
+        }
+
+    def _forbidden_code(self):
+        return "(commit|rollback)"
```
Comment (jarroyomorales) [inline kpi_dashboard/models/kpi_kpi.py]: Shouldnt you also forbid `execute` ?

**Review by bealdav — APPROVED** [review]

**Comment by bealdav** [comment]
Hi @legalsylvain your wishes are anwsered. Thanks

File: kpi_dashboard/models/kpi_kpi.py (hunk 3 above)
Comment (legalsylvain) [reply kpi_dashboard/models/kpi_kpi.py]: Well, in that case, it will avoid to have a compute value code that makes an SQL query, that can be interesting.

Comment (etobella) [reply kpi_dashboard/models/kpi_kpi.py]: Good Point, I have added execute to the forbidden words. I also added `getattr`, as you could do something like:
```
getattr(self.env.cr, 'exec'+'ute')('com'+'mit')
```

Comment (pedrobaeza) [reply kpi_dashboard/models/kpi_kpi.py]: But it's the contrary. @legalsylvain is saying that this prevent manual SQL queries, and the only problem I see is to commit that, so no need to forbid it.

File: kpi_dashboard/models/kpi_kpi.py [hunk 4]
```diff
@@ -84,6 +88,27 @@ def write(self, vals):
...
+            "model": self.browse(),
+        }
+
+    def _forbidden_code(self):
+        return "(commit|rollback|getattr|execute)"
+
+    def _compute_value_code(self):
+        if re.search(self._forbidden_code(), (self.code or "").lower()):
+            raise ValidationError(_(
+                "The code cannot contain rollback or commit inside."
```
Comment (bealdav) [inline kpi_dashboard/models/kpi_kpi.py]: Maybe 
```suggestion
                "The code cannot contain these terms '%s' inside." % self._forbidden_code()[1:-1].replace("|", ", ")
```

Comment (etobella) [reply kpi_dashboard/models/kpi_kpi.py]: I proposed another option. What do you think @bealdav ?

**Comment by LoisRForgeFlow** [comment]
Then, do you think guys this is good to merge? We shouldn't stop momentum ;)

**Comment by bealdav** [comment]
Yes I think it's good to merge. It doesn't prevent to improve it after

**Comment by pedrobaeza** [comment]
Last change is not going to be in the merge. Should we stop the merge process for including it?

**Comment by etobella** [comment]
I was adding myself as maintainer

**Comment by LoisRForgeFlow** [comment]
cancelled, let's restart:

/ocabot merge minor

**Comment by pedrobaeza** [comment]
Other thing: this topic about avoiding to abuse the `safe_eval` thing is in several modules (base_tier_validation_formula - or similar name -, etc). Maybe we should extract this logic in a base module that join all the needed techniques and make use of it. It can be named `base_super_safe_eval` :stuck_out_tongue_closed_eyes:

**Comment by etobella** [comment]
Maybe it is interesting to create a module on server-tools or somewhere else. If we create the module I will integrate it here. What do you think @LoisRForgeFlow ?

**Comment by LoisRForgeFlow** [comment]
Sounds good, I like security improvements :nerd_face:

**Pull request merged**